### PR TITLE
Feat/fees management

### LINKED ISSUE #199: Define Behaviour for FeesManager and DepositCalculator traits

### PR DESCRIPTION
**Pull request by driemworks**
Closes #199 
This implements a simplified pricing model that:
- allows for a configurable base fee
- sets the idle credits and dispatch credits both to 1
- uses the tiered pricing schedule
- verifies economic assumptions with zombienet

### CODE REVIEW

File: e2e/tests/subscriptions/test-economics.zndsl [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+# expected credits: 1
+# expected fees: 900_000
+# expected deposit: 1020
+# expected subscription duration: 1 block = 6 secs
+###############################################################
+# SETUP
+########################################################
+# zero-out the treasury balance
+idn-alice: js-script ../../scripts/testing/utils/reset-balance.js with "5CQE1RtAnMdcdWgx4EuvnGYfdPa5qwQS2pQMzhjsPn7k3A1C" within 120 secs
+idn-alice: js-script ../../scripts/testing/utils/check-balance.js with "5CQE1RtAnMdcdWgx4EuvnGYfdPa5qwQS2pQMzhjsPn7k3A1C, 1000000" within 120 secs
```
Comment (juangirini) [inline e2e/tests/subscriptions/test-economics.zndsl:38]: either the tags or title are wrong, idn-alice is a collator, not treasury. Also, if it was the treasury the tests shouldn't pass, as the treasury needs the minimum existential deposit in order to be able to receive tokens

File: pallets/idn-manager/src/lib.rs [hunk 2]
```diff
@@ -514,6 +514,8 @@ pub mod pallet {
 			// Calculate and hold the subscription fees
 			let fees = Self::calculate_subscription_fees(&params.credits);
 
+			// panic!("{:?}", fees);
+
```
Comment (juangirini) [inline pallets/idn-manager/src/lib.rs:518]:
```suggestion
```

**Review by juangirini — APPROVED** [review]
Looks good, there's one thing I don't understand in the tests. Left you a comment

File: e2e/tests/subscriptions/test-economics.zndsl (hunk 1 above)
Comment (driemworks) [reply e2e/tests/subscriptions/test-economics.zndsl:38]: The comment is wrong, will update. The original plan was to 'zero' out the balance, but that isn't a possibility, so we reduce it to the existential deposit it instead.

File: pallets/idn-manager/src/lib.rs (hunk 2 above)
Comment (driemworks) [reply pallets/idn-manager/src/lib.rs]: removed

File: e2e/tests/subscriptions/test-economics.zndsl [hunk 3]
```diff
@@ -33,17 +34,17 @@ sleep 7 seconds
 ###############################################################
 # SETUP
 ########################################################
-# zero-out the treasury balance
+# set the treasury balance to existential deposit (1_000_000)
 idn-alice: js-script ../../scripts/testing/utils/reset-balance.js with "5CQE1RtAnMdcdWgx4EuvnGYfdPa5qwQS2pQMzhjsPn7k3A1C" within 120 secs
-idn-alice: js-script ../../scripts/testing/utils/check-balance.js with "5CQE1RtAnMdcdWgx4EuvnGYfdPa5qwQS2pQMzhjsPn7k3A1C, 1000000" within 120 secs
+#idn-alice: js-script ../../scripts/testing/utils/check-balance.js with "5CQE1RtAnMdcdWgx4EuvnGYfdPa5qwQS2pQMzhjsPn7k3A1C, 1000000" within 120 secs
```
[the remark this answers is not in the document]
Comment (driemworks) [reply e2e/tests/subscriptions/test-economics.zndsl]: removed

File: e2e/tests/subscriptions/test-economics.zndsl [hunk 4]
```diff
@@ -68,17 +69,17 @@ idn-alice: js-script ../../scripts/testing/utils/check-balance.js with "5CQE1RtA
 ######################################################
 # SETUP
 ########################################################
-# zero-out the treasury balance
+# set the treasury balance to existential deposit (1_000_000)
 idn-alice: js-script ../../scripts/testing/utils/reset-balance.js with "5CQE1RtAnMdcdWgx4EuvnGYfdPa5qwQS2pQMzhjsPn7k3A1C" within 120 secs
-idn-alice: js-script ../../scripts/testing/utils/check-balance.js with "5CQE1RtAnMdcdWgx4EuvnGYfdPa5qwQS2pQMzhjsPn7k3A1C, 1000000" within 120 secs
+#idn-alice: js-script ../../scripts/testing/utils/check-balance.js with "5CQE1RtAnMdcdWgx4EuvnGYfdPa5qwQS2pQMzhjsPn7k3A1C, 1000000" within 120 secs
```
[the remark this answers is not in the document]
Comment (driemworks) [reply e2e/tests/subscriptions/test-economics.zndsl]: removed

File: chains/idn-consumer-chain/runtime/src/genesis_config_presets.rs [hunk 5]
```diff
@@ -88,7 +88,7 @@ fn local_testnet_genesis() -> Value {
 		],
 		Sr25519Keyring::well_known().map(|k| k.to_account_id()).collect(),
 		Sr25519Keyring::Alice.to_account_id(),
-		2000.into(),
+		4502.into(),
```
Comment (colemanirby) [inline chains/idn-consumer-chain/runtime/src/genesis_config_presets.rs:91]: I believe that this should still be 2000.into(). If zombienet related functionality isn't failing then maybe it's fine, but 4502 is the paraid of the IDN and 2000 is the paraid of the consumer chain (in zombienet)

File: chains/idn-consumer-chain/runtime/src/genesis_config_presets.rs [hunk 6]
```diff
@@ -101,7 +101,7 @@ fn development_config_genesis() -> Value {
 		],
 		Sr25519Keyring::well_known().map(|k| k.to_account_id()).collect(),
 		Sr25519Keyring::Alice.public().into(),
-		2000.into(),
```
Comment (colemanirby) [inline chains/idn-consumer-chain/runtime/src/genesis_config_presets.rs:104]: Same as previous comment

File: chains/idn-consumer-chain/runtime/src/genesis_config_presets.rs (hunk 5 above)
Comment (colemanirby) [reply chains/idn-consumer-chain/runtime/src/genesis_config_presets.rs:91]: In the paseo-local.json the relay chain will be looking for a parachain with paraid 4502. Checking the zombienet.toml, I see that we have the paraid of the consumer chain to be 2001. This is the same as the main branch, so I suppose having a different paraid here isn't breaking anything (since on main we specify 2000 in this file, but also specify 2001 as paraid in the zombienet.toml file)

File: pallets/idn-manager/src/impls.rs [hunk 7]
```diff
@@ -259,12 +258,12 @@ where
 
 	/// Get the number of credits consumed by a subscription when this one gets a pulse in a block.
 	fn get_consume_credits(_sub: Option<&Sub>) -> u64 {
-		1000
+		1
```
Comment (colemanirby) [inline pallets/idn-manager/src/impls.rs:261]: I just want to be sure, we're going to change this to be a config value in a different PR right?

**Review by colemanirby — APPROVED** [review]
LGTM

File: pallets/idn-manager/src/impls.rs (hunk 7 above)
Comment (driemworks) [reply pallets/idn-manager/src/impls.rs:261]: https://github.com/ideal-lab5/idn-sdk/issues/311
